virttest/qemu_devices/qdevices.py

```diff
@@ -1900,6 +1909,24 @@ def get_free_root_port(self):
 
         return None
 
+    def prepare_hotplug(self, device):
+        """
+        Prepares a device to be hot-plugged into this bus;
+        You need to call "insert" afterwards. Concurrent calls
+        of "prepare_hotplug" and "insert" are not supported
+        :param device: The QDevice object
+        """
+        root_port = self.get_free_root_port()
+        device.set_param("bus", root_port)
+        if isinstance(device.parent_bus, tuple):
+            device.parent_bus = list(device.parent_bus)
+        elif not isinstance(device.parent_bus, list):
+            device.parent_bus = [device.parent_bus]
+        for bus in device.parent_bus:
+            if self.match_bus(bus):
+                device.parent_bus.remove(bus)
```

Comment:
iterating through list and removing objects is insecure (it can skip items). Try this:

```python
buses = ["pci.0", "bar", "pci.0", "pci.0"]
for bus in buses:
    if bus == 'pci.0':
        buses.remove(bus)

print(buses)
```

Something like this would work.

```python
remove = [bus for bus in device.parent_bus if self.match_bus(bus)]
for bus in remove:
    device.parent_bus.remove(bus)
```

But given the parent_bus is frequently tuple, how about:

```
root_port = self.get_free_root_port()
device.set_param("bus", root_port)
if not isinstance(device.parent_bus, (tuple, list)):
    device.parent_bus = [device.parent_bus]
device.parent_bus = tuple(bus for bus in device.parent_bus
                          if not self.match_bus(bus)) + ({"busid": root_port},)
```

which avoids the tuple->list conversion?